activesupport/lib/active_support/core_ext/enumerable.rb

```diff
@@ -191,6 +195,28 @@ def pick(*keys)
   def compact_blank
     reject(&:blank?)
   end
+
+  # Returns the sole item in the enumerable. If there are no items, or more
+  # than one item, raises +Enumerable::SoleItemExpectedError+.
+  #
+  #    ["x"].sole          # => "x"
+  #    Set.new.sole        # => Enumerable::SoleItemExpectedError: no item found
+  #    { a: 1, b: 2 }.sole # => Enumerable::SoleItemExpectedError: multiple items found
+  def sole
+    it = nil
+    found = false
```

Comment:
Can't we use the same technique of `first(2)`?

Reply:
I've made this change, but realised it doesn't let us distinguish these cases:

```ruby
irb(main):001:0> found, undesired = [:a].first(2)
=> [:a]
irb(main):002:0> found
=> :a
irb(main):003:0> undesired
=> nil
irb(main):004:0> found, undesired = [:a, nil].first(2)
=> [:a, nil]
irb(main):005:0> found
=> :a
irb(main):006:0> undesired
=> nil
irb(main):007:0> 
```

As it stands, `sole` will now not raise if an Enumerable yields two items, the second of which is `nil`.

The original change per https://github.com/rails/rails/pull/26206#issuecomment-241208216 does distinguish this case, so I think I prefer it.

Reply:
I've reverted the change of algorithm in 0c58f6f817.

Reply:
That strikes me as more confusing, to be honest. That an extra element that happens to be nil isn't considered an element. I'd rather we treat nil as an element in this evaluation. So I think we should revert to the first(2) approach.

Reply:
> I'd rather we treat nil as an element in this evaluation. So I think we should revert to the first(2) approach.

I'm a bit confused by this comment.

I've reverted to the `first(2)` approach as requested, but — per the example irb session above — a second item which happens to be `nil` is not distinguished from no second item at all, as `x, y = first(2)` will put a `nil` into the `y` in either case. As a result, `[:a, nil].sole == :a`.

The alternative implementation (kivikakk@a7c90384c77) causes a `SoleItemExpectedError` in this case. I think that's more correct and less likely to cause unexpected errors, but happy to defer to you here.

Reply:
Ahh, my apologies. I misread that. Hmm, yeah, I think you're right. An array that has 2 elements, one not-nil and one nil, should indeed error out. Good to proceed with your alternative implementation 👍

Reply:
That's been pushed to this branch, along with an assertion in the test for that behaviour. 👍

Reply:
Appreciate all the ping ponging on this. I think what we have now is totally workable and fine. But aesthetically, I don't love the multiple exit paths across different scopes. So one last suggestion before we commit. How about this?

```ruby
case count
when 1   then return first
when 0   then raise SoleItemExpectedError, "no item found"
when 2.. then raise SoleItemExpectedError, "multiple items found"
end
```

And again, really appreciate your patience to get this done ✌️❤️

Reply:
Yes, this is preferable — I don't love the number of guard clauses in the old implementation anyway, it's difficult to trace the flow.

The possible downside to doing it this way is if the underlying Enumerable's `count` does something expensive. It could just as easily do something expensive to enumerate with `each`, so I don't consider that a stopper.

Reply:
c25dd0a makes this change.

Reply:
@kivikakk Yeah, hopefully even if #count is doing something expensive, you're still calling this #sole with the expectation that the set if very small. So even O(n) or worse presumably should be cheap! It's an exceptional state if there are even 2 elements.

Did Rubocop reject using an explicit "return" for the first case? If so, great example for why I often end up strongly disliking Rubocop 😂. Normally, I'd agree that adding return is not something to have. But in this case, since it's being contrasted with "raise", I liked it better being explicit.

Reply:
Yeah, it did 😅 <https://github.com/rails/rails/runs/2379000836?check_suite_focus=true>

```
activesupport/lib/active_support/core_ext/enumerable.rb:231:19: C: [Correctable] Style/RedundantReturn: Redundant return detected.
    when 1   then return first
                  ^^^^^^

2912 files inspected, 1 offense detected, 1 offense auto-correctable
Error: Process completed with exit code 1.
```

I think it looks better with the `return` too. It looks like we don't use per-line Rubocop disables very much, though?

Reply:
Let's make a point to do show that rebelling against Rubocop is possible ✊. So please do add that per-line exemption, and let's put the return back in there. It's this level of finesse and subtlety that brings charm and joy to writing Ruby 😄❤️

Reply:
Awesome. Done!

Reply:
Can we use `size` instead of `count`?
Unlike `size`, `count` on a relation will do a COUNT query even if a relation is loaded.

Reply:
AR already has its own implementation of #sole. This is for other Enumerables.

Reply:
Ah I didn't know that (#40768), thank you for letting me know!